cupy/cuda/function.pyx

```diff
@@ -26,51 +26,13 @@ cdef class CPointer:
     def __init__(self, p=0):
         self.ptr = <void*>p
 
-
-cdef class CInt8(CPointer):
-    cdef:
-        int8_t val
-
-    def __init__(self, int8_t v):
-        self.val = v
-        self.ptr = <void*>&self.val
-
-
-cdef class CInt16(CPointer):
-    cdef:
-        int16_t val
-
-    def __init__(self, int16_t v):
-        self.val = v
-        self.ptr = <void*>&self.val
-
-
-cdef class CInt32(CPointer):
+cdef class CNumpyArray(CPointer):
     cdef:
-        int32_t val
-
-    def __init__(self, int32_t v):
-        self.val = v
-        self.ptr = <void*>&self.val
-
-
-cdef class CInt64(CPointer):
-    cdef:
-        int64_t val
-
-    def __init__(self, int64_t v):
-        self.val = v
-        self.ptr = <void*>&self.val
-
-
-cdef class CInt128(CPointer):
-    cdef:
-        double complex val
-
-    def __init__(self, double complex v):
-        self.val = v
-        self.ptr = <void*>&self.val
+        int8_t[:] val
 
+    def __init__(self, a):
+        self.val = numpy.atleast_1d(a).ravel().view(numpy.int8)
```

Comment:
We are a bit concerned about this,
we want to have kernel launch as fast as possible and calling several full python numpy routine for every argument that we pass can be overkill and bump the launch time from a few microseconds to several hundreds.

I still have to measure this, I expect to do it tomorrow. In the case that the overhead is significant, the solution here will be to leave the old values as they are and just add this one when it is necessary. But this will need some reworking to deal with the array pointer instead of all these calls.

Reply:
I am not an expert in cython at all and I don't think val is required here. I think you can directly use `val.ctypes.data`  without defining the memory view but I dit not manage to cast it correctly with the `<void*>` syntax.

On my side `python3 -m timeit -s 'import numpy as np; a=np.float64(12);' -- 'np.atleast_1d(a).ravel().view(np.int8)'` gives me 2.03 usec per loop which is effectively a cost that we should get rid off.